### PR TITLE
Add support for faster whitespace tolerant judge

### LINKED ISSUE #173: Tokenizer
Diff has quadratic complexity. Better replace it with something linear.

### PR DESCRIPTION
**Pull request by SvizelPritula**
This PR imports KSP's `judgelib` judge utility library along with the `judge-token` program. It also adds support for using it instead of `diff`.

Both `out_check=diff` and `out_check=tokens` are supported. If we wan't to deprecate `diff` or alias it to `tokens`, we can easily do so in the future. It also supports all options supported by KSP's opendata tool.

One slight difference between this and KSP's implementation is that the floating point mode is disabled by setting `tokens_float_rel_error` and `tokens_float_abs_error` to `X` instead of the empty string.

Closes #173.

### CODE REVIEW

File: config-v3-documentation [hunk 1]
```diff
@@ -62,6 +62,7 @@ checker=check
 out_check=judge
 # Describes how to check outputs
 #   - diff: checks that contestant output equals correct output (ignores whitespace) (default)
+#   - tokens: uses a fast, versatile file equality checker (ignores whitespace)
```
Comment (5K1PY) [inline config-v3-documentation:65]: Isn't tokens a better default?

File: config-v3-documentation [hunk 2]
```diff
@@ -85,6 +86,25 @@ judge_needs_out=0
...
+# If set to 1, ASCII characters will be compared in a case-insensitive manner
+
+tokens_float_rel_error=0.00001
+tokens_float_abs_error=1e-30
+# Only for out_check=tokens
+# When these options are specified, floating-point numbers
+# will be parsed and compared with a given error margin
+# Any tokens that can't be parsed as a float will be compared character-by-character
+# If used, both of these options must be specified
+# Set both options to X to disable
```
Comment (5K1PY) [inline config-v3-documentation:106]: I don't really know about the `X`. We use it in different context with a different meaning. I would much prefer either `@` value or empty to disable.

File: pisek/config/task_config.py [hunk 3]
```diff
@@ -226,6 +237,18 @@ def validate_model(self):
                 {"task_type": self.task_type, "out_check": self.out_check},
             )
 
+        if (self.tokens_float_abs_error == None) != (
```
Comment (5K1PY) [inline pisek/config/task_config.py:240]: Nit: Pyhton style prefers ``is None`` instead of ``== None``.

File: config-v3-documentation (hunk 1 above)
Comment (SvizelPritula) [reply config-v3-documentation:65]: Probably. I wanted to keep the default unchanged for backwards compatibility, but that might not be necessary. There aren't any Kasiopea tasks that doesn't set `out_check` explicitly.

File: pisek/jobs/parts/tools.py [hunk 4]
```diff
@@ -38,6 +38,7 @@ def _get_jobs(self) -> list[Job]:
         jobs: list[Job] = [
             PrepareMinibox(self._env),
             PrepareTextPreprocessor(self._env),
+            PrepareJudgeToken(self._env),
```
Comment (5K1PY) [inline pisek/jobs/parts/tools.py:41]: Do we really want to always compile this?

File: pisek/jobs/parts/tools.py [hunk 5]
```diff
@@ -100,6 +101,41 @@ def _run(self):
             raise PipelineItemFailure("Text preprocessor compilation failed.")
 
 
+class PrepareJudgeToken(TaskJob):
```
Comment (5K1PY) [inline pisek/jobs/parts/tools.py:104]: Nit: I would call it ``PrepareTokenJudge`` to have same name

Comment (5K1PY) [inline pisek/jobs/parts/tools.py:104]: Nit: I would call it ``PrepareTokenJudge`` to have same name as

Comment (5K1PY) [inline pisek/jobs/parts/tools.py:104]: Nit: I would call it ``PrepareTokenJudge`` to have same name as ``RunTokenJudge``.

**Review by 5K1PY — CHANGES_REQUESTED** [review]

File: config-v3-documentation (hunk 1 above)
Comment (5K1PY) [reply config-v3-documentation:65]: I was thinking about adding it to update script

Comment (SvizelPritula) [reply config-v3-documentation:65]: What do you want to add to the update script? Adding `out_check=diff` if `out_check` isn't specified?

Comment (5K1PY) [reply config-v3-documentation:65]: Yup

File: pisek/jobs/parts/tools.py (hunk 5 above)
Comment (SvizelPritula) [reply pisek/jobs/parts/tools.py]: I was unsure if I should use the word order used by the program `judge-token` or the more grammatical order. Apparently, I chose both. Fixed.

File: pisek/jobs/parts/tools.py (hunk 4 above)
Comment (SvizelPritula) [reply pisek/jobs/parts/tools.py]: Fixed. The judge is now compiled on demand.

File: pisek/config/task_config.py (hunk 3 above)
Comment (SvizelPritula) [reply pisek/config/task_config.py]: Fixed.

File: config-v3-documentation (hunk 2 above)
Comment (SvizelPritula) [reply config-v3-documentation:106]: I thought it was kind of similar to `max_submissions`. I've changed the unset value to the empty string.

File: config-v3-documentation (hunk 1 above)
Comment (SvizelPritula) [reply config-v3-documentation]: Turns out the documentation was wrong: The `out_check` key has no default and is mandatory. I've fixed the documentation.

File: config-v3-documentation (hunk 2 above)
Comment (5K1PY) [reply config-v3-documentation:106]: > I thought it was kind of similar to `max_submissions`. I've changed the unset value to the empty string.

Something like `@unlimited` could be better, but that's up to another PR.

Comment (5K1PY) [reply config-v3-documentation:106]: > I thought it was kind of similar to `max_submissions`. I've changed the unset value to the empty string.

Can you also update the documentation?

Comment (SvizelPritula) [reply config-v3-documentation]: Fixed.

**Review by 5K1PY — APPROVED** [review]
LGTM. Great work!

**Pull request merged**